inlong-sort/sort-formats/format-inlongmsg-pb/src/main/java/org/apache/inlong/sort/formats/inlongmsgpb/InLongMsgPbDecodingFormat.java

```diff
@@ -0,0 +1,237 @@
+ * InLongMsg pb format decoding format.
+ */
+public class InLongMsgPbDecodingFormat implements DecodingFormat<DeserializationSchema<RowData>> {
+
+    private static final Logger log = LoggerFactory.getLogger(InLongMsgPbDecodingFormat.class);
+
+    private final String innerFormatMetaPrefix;
+
+    private final DecodingFormat<DeserializationSchema<RowData>> innerDecodingFormat;
+
+    private List<String> metadataKeys;
+
+    private final boolean ignoreErrors;
+
+    private final String decompressType;
+
+    private final boolean ignoreTailingUnmappable;
+
+    public InLongMsgPbDecodingFormat(
+            DecodingFormat<DeserializationSchema<RowData>> innerDecodingFormat,
+            String innerFormatMetaPrefix,
+            boolean ignoreErrors,
+            boolean ignoreTailingUnmappable,
+            String decompressType) {
+        this.innerDecodingFormat = innerDecodingFormat;
+        this.innerFormatMetaPrefix =  innerFormatMetaPrefix;
+        this.metadataKeys = Collections.emptyList();
+        this.ignoreErrors = ignoreErrors;
+        this.ignoreTailingUnmappable = ignoreTailingUnmappable;
+        this.decompressType = decompressType;
+    }
+
+    @Override
+    public DeserializationSchema<RowData> createRuntimeDecoder(Context context, DataType physicalDataType) {
+        final MetadataConverter[] metadataConverters = Arrays.stream(ReadableMetadata.values())
+                .filter(metadata -> metadataKeys.contains(metadata.key))
+                .map(metadata -> metadata.converter)
+                .toArray(MetadataConverter[]::new);
+        final List<ReadableMetadata> readableMetadata =
+                metadataKeys.stream()
+                        .map(
+                                k ->
+                                        Stream.of(ReadableMetadata.values())
+                                                .filter(rm -> rm.key.equals(k))
+                                                .findFirst()
+                                                .orElseThrow(IllegalStateException::new))
+                        .collect(Collectors.toList());
+        final List<DataTypes.Field> metadataFields =
+                readableMetadata.stream()
+                        .map(m -> DataTypes.FIELD(m.key, m.dataType))
+                        .collect(Collectors.toList());
+        final DataType producedDataType =
+                DataTypeUtils.appendRowFields(physicalDataType, metadataFields);
+        final TypeInformation<RowData> producedTypeInfo =
+                context.createTypeInformation(producedDataType);
+        final InLongPbMsgDecompressor decompressor = getDecompressor(decompressType);
+
+        DeserializationSchema<RowData> innerSchema =
+                innerDecodingFormat.createRuntimeDecoder(context, physicalDataType);
+        if (innerSchema instanceof CsvRowDataDeserializationSchema && ignoreTailingUnmappable) {
+            this.makeCsvInnerFormatIgnoreTailingUnmappable(innerSchema);
```

Comment:
why here a special treatment？could you describe it more clearly in this [4807](https://github.com/apache/inlong/issues/4807)

Reply:
Better to give an example

Reply:
note has been added, ptal at #4807